### PR TITLE
Allow changing output names for static libs and install pdb files for static archives.

### PR DESCRIPTION
**Pull request by stevewgr**
### Description

- [x] Read CONTRIBUTING.md, a link to which appears under "Helpful resources" below.  That document discusses general guidelines for contributing to libjpeg-turbo, as well as the types of contributions that will not be accepted or are unlikely to be accepted.
- [x] Search the existing issues and pull requests (both open and closed) to ensure that a similar request has not already been submitted and rejected.
- [x] Discuss the proposed bug fix or feature in a GitHub issue, through direct e-mail with the project maintainer, or on the libjpeg-turbo-devel mailing list.

Note that on msvc it also generates pdb files for static archives. Surely we can embed them with /Z7 linker flag, however it is often a case where one also distributes their static libraries with their coressponding pdb's.
There are open open tickets on cmake end to implement `TARGET_COMPILE_PDB_FILE` for this use case, however that will likely not happen anytime soon:
- https://gitlab.kitware.com/cmake/cmake/-/issues/16932
- https://gitlab.kitware.com/cmake/cmake/-/issues/16935

Therefore the approach I added should work on all scenarios.

As for renaming the static archives output names, one could provide a cache argument to cmake via `-DJPEG_STATIC_OUTPUT_NAME ` and `-DTURBOJPEG_STATIC_OUTPUT_NAME`. Why is this useful? Because I don't want to build the shared libraries, so I'd rather distribute the build from your project containing only the static libraries. Note that the code I added for this behavior does not change the existing behavior so that it is backward compatible with previous use cases. Therefore one can completely ignore them when configuring cmake.

### CODE REVIEW

**Comment by dcommander** [comment]
You clicked through the PR prerequisites without actually performing any of those steps. (I am very adamant about the fact that I want to discuss any changes before a PR is  submitted, because about 99% of PRs I receive fundamentally misconstrue something about how libjpeg-turbo is supposed to work.) These patches appear to only benefit your specific downstream build environment. libjpeg-turbo is used indirectly by billions of people every day (through popular web browsers) and is an ISO/ITU-T reference implementation. I'm not sure how many developers consume it downstream, but I would guess tens, if not hundreds, of thousands. Thus, it cannot be a repository for the individual preferences of everyone who uses it. That would cause it to become an unmaintainable mess.

The **.gitignore** patch is certainly not applicable or acceptable upstream. If you want to ignore specific build directories in your downstream build, then put those in **.git/info/exclude**. As to the other patches, please justify how those would benefit the general case and not just your specific case. If the goal is simply to remove the -static suffix from the libraries, then can't you already do that by setting `ENABLE_SHARED=0` in your build? Your description is unclear as to why you need to rename the libraries.

**Comment by stevewgr** [comment]
> You clicked through the PR prerequisites without actually performing any of those steps. (I am very adamant about the fact that I want to discuss any changes before a PR is submitted, because about 99% of PRs I receive fundamentally misconstrue something about how libjpeg-turbo is supposed to work.)

Oh no! I'm so sorry 😱 I read through it ✅ and will remove the `.gitignore` patch.

> These patches appear to only benefit your specific downstream build environment.

I politely disagree, because as a library author it makes sense giving the developers the flexibility to make this project adaptable to their own project needs. The changes introduced here do not influence the existing behavior, rather fixing missing steps, such as installing pdb files for static libraries and msvc inconsistencies with other platforms using `*.lib` extension for static libraries.

> The **.gitignore** patch is certainly not applicable or acceptable upstream. If you want to ignore specific build directories in your downstream build, then put those in **.git/info/exclude**.

Yes, I guess I know how to use `git`, but thanks nevertheless. 😕

> As to the other patches, please justify how those would benefit the general case and not just your specific case. If the goal is simply to remove the -static suffix from the libraries, then can't you already do that by setting `ENABLE_SHARED=0` in your build? Your description is unclear as to why you need to rename the libraries.

Yep, I already use `ENABLE_SHARED=OFF` in my build, but that has nothing to do with the output name of the static library. Here is an example of common configurations I often use when building cmake based projects, just focused on your project:
```
cmake .. -G "Visual Studio 17 2022" -A Win32 -DENABLE_SHARED=OFF -DWITH_TURBOJPEG=OFF -DJPEG_STATIC_OUTPUT_NAME=jpeg -DCMAKE_INSTALL_PREFIX=pkg -DCMAKE_MSVC_DEBUG_INFORMATION_FORMAT="$<$<CONFIG:Debug,RelWithDebInfo>:Embedded>" -DCMAKE_POLICY_DEFAULT_CMP0141=NEW

cmake --build . --config Debug --target jpeg-static install
cmake --build . --config Release --target jpeg-static install
```

**Comment by stevewgr** [comment]
@dcommander I dropped the `.gitignore` commit.

**Comment by stevewgr** [comment]
>  Discuss the proposed bug fix or feature in a GitHub issue, through direct e-mail with the project maintainer, or on the libjpeg-turbo-devel mailing list.

considering we're already discussing it in the PR, do you still need me to create an issue for it?

**Comment by dcommander** [comment]
> considering we're already discussing it in the PR, do you still need me to create an issue for it?

No. It is just a red flag whenever people click through the prerequisites. It makes me immediately suspicious of the PR. Maintaining this project requires an extreme amount of attention to detail, because it is complicated and ubiquitous and cross-platform (to an extreme level) and has a lot of old code that takes a lot of time to fully understand. That is why it isn't a community-developed project. If GitHub allowed me to disable the PR system for this project or to require that an issue be created first, I would do that. Unsolicited PRs are simply not a good fit for our development model. But what's done is done, so we can keep discussing it here.

**Comment by dcommander** [comment]
> I politely disagree, because as a library author it makes sense giving the developers the flexibility to make this project adaptable to their own project needs. The changes introduced here do not influence the existing behavior, rather fixing missing steps, such as installing pdb files for static libraries and msvc inconsistencies with other platforms using `*.lib` extension for static libraries.

A project such as this must be maintained with an eye toward minimal intrusiveness. No one else has asked for this change, so I need to understand what your ultimate goal is, why that goal cannot currently be achieved with libjpeg-turbo's build system, and how to minimally achieve that goal. See below for more specifics.

> > The **.gitignore** patch is certainly not applicable or acceptable upstream. If you want to ignore specific build directories in your downstream build, then put those in **.git/info/exclude**.
> 
> 
> 
> Yes, I guess I know how to use `git`, but thanks nevertheless. 😕

Submitting a patch that binds our upstream project to your specific downstream choice of build directory nomenclature, when Git provides a way of achieving the same result locally, doesn't inspire confidence.

> Yep, I already use `ENABLE_SHARED=OFF` in my build, but that has nothing to do with the output name of the static library.

But the build system is supposed to remove the -static suffixes when `ENABLE_SHARED` is off. So I'm trying to figure out whether (a) that is your ultimate goal and (b) whether there is a bug in the build system that is preventing it.

**Comment by stevewgr** [comment]
> > I politely disagree, because as a library author it makes sense giving the developers the flexibility to make this project adaptable to their own project needs. The changes introduced here do not influence the existing behavior, rather fixing missing steps, such as installing pdb files for static libraries and msvc inconsistencies with other platforms using `*.lib` extension for static libraries.
> 
> A project such as this must be maintained with an eye toward minimal intrusiveness. No one else has asked for this change, so I need to understand what your ultimate goal is, why that goal cannot currently be achieved with libjpeg-turbo's build system, and how to minimally achieve that goal. See below for more specifics.


> > Yep, I already use `ENABLE_SHARED=OFF` in my build, but that has nothing to do with the output name of the static library.
> 
> But the build system is supposed to remove the -static suffixes when `ENABLE_SHARED` is off. So I'm trying to figure out whether (a) that is your ultimate goal and (b) whether there is a bug in the build system that is preventing it.

It is really simple as it is: you're not setting the output name of your static libraries unless it is not msvc: https://github.com/libjpeg-turbo/libjpeg-turbo/blob/3c17063ef1ab43f5877f19d670dc39497c5cd036/CMakeLists.txt#L667-L672

The default behavior of cmake here is to set the output name the same as the target name, which brings us back to the initial topic I mentioned before; on Unix systems it is `*.a`, which prevents name collisions. However on Windows static archives are also `*.lib`, same as with dynamic libs. Therefore the check made there is not a bug, rather a feature, but enforces weird behavior on Windows users consuming your library.

I swear I'm not reinventing the wheel here lol

**Comment by stevewgr** [comment]
> But the build system is supposed to remove the -static suffixes when `ENABLE_SHARED` is off.

Also allow me to correct you specifically on this: there is no such behavior in your cmake scripts. I mean there is no dependecy on the output name of static libraries based on `ENABLED_SHARED`, they're independent from one another.

**Comment by dcommander** [comment]
I am out of the office, so I will have to look at this more closely next week and figure out how best to address it in a way that avoids unintended consequences. Allowing the library name to be changed arbitrarily would break the packaging system, and there may be other points of breakage as well. I am also uncomfortable with allowing any arbitrary name for our library, since that introduces an additional support vector. (If downstream projects want to rename our library to **foo.lib**, then that's their business, but it isn't my job to deal with the fallout from that. Adding the variable to the upstream build system would imply that it is my job.)

I am much more comfortable with simply removing the **-static** suffix from the Visual C++ libraries if `ENABLE_SHARED` is off. However, that is also potentially a breaking change, since other downstream projects may rely on the existing nomenclature (and our own documentation specifies that nomenclature, regardless of the value of `ENABLE_SHARED`.) Trust me. No matter how simple you think the change is, I guarantee you that someone will complain about it. No good deed ever goes unpunished with this project, which is why I have to be ultra careful about modifying it. (But that only reduces the number of complaints. It doesn't eliminate them.) For better or worse, because so many downstream projects use libjpeg-turbo (and libjpeg before it) in "creative" ways, every nuance of the library's behavior and the interfaces it provides have essentially become de facto standards. If I change something that may require a downstream project to change something, then I minimally need to be prepared to deliver a full-throated justification of the change.

I need to look at how other projects are handling the same issue. A new CMake variable may be required, but it would be a lot easier to support a binary variable than a variable with an arbitrary value.

**Comment by stevewgr** [comment]
> Allowing the library name to be changed arbitrarily would break the packaging system, and there may be other points of breakage as well. I am also uncomfortable with allowing any arbitrary name for our library, since that introduces an additional support vector. (If downstream projects want to rename our library to **foo.lib**, then that's their business, but it isn't my job to deal with the fallout from that. Adding the variable to the upstream build system would imply that it is my job.)

I believe maybe then there is a misunderstanding here with the changes introduced and [how cmake cache works](https://cmake.org/cmake/help/book/mastering-cmake/chapter/CMake%20Cache.html). The changes introduced in this PR are not breaking changes and does not change how your current cmake files works, rather introducing more flexibility due to unecessary restriction in your scripts for Windows users enforcing them to use the `-static` suffix for a static library, whilst on other platforms that's not the case. If you're concerned about downstream projects renaming your library, then you should close source it, really ;) I can fully manipulate your build from the official release tag by using cmake cache variables. One example where I embed the symbols instead of generating pdbs during the config stage (`-DCMAKE_MSVC_DEBUG_INFORMATION_FORMAT="$<$<CONFIG:Debug,RelWithDebInfo>:Embedded>"`) or enforcing it to use Microsoft runtime DLLs (`-DCMAKE_MSVC_RUNTIME_LIBRARY_RELEASE="MultiThreadedDLL"`). What I try to say here is that even if you set these variables in your scripts, one can already cache them via command line arguments, just as an example.
So my approach in this PR was: "hey, you want to set the output name for static libraries? go ahead my friend, otherwise we'll just use the default output names as before". Why? because static libraries anyways get embedded into the final build in comparison to dynamic libraries where one have to distribute your dlls with their project. So why you're worried about other folks renaming the static libraries too much, I'm honestly not sure.

Ignoring the fact that it also fixing issues with your scripts not installing symbols for static libraries AND actually providing the postfix when setting the output name, depending whether it is release or debug builds. With your current implementation both Release and Debug for non msvc builds, will produce the same postfix, meaning you can't distinguise between the two builds 💥 So two very good positive fixes 😎 

> I am much more comfortable with simply removing the **-static** suffix from the Visual C++ libraries if `ENABLE_SHARED` is off. However, that is also potentially a breaking change, since other downstream projects may rely on the existing nomenclature (and our own documentation specifies that nomenclature, regardless of the value of `ENABLE_SHARED`.) Trust me.

Well, what you propose here you named it yourself, this is indeed a breaking change and hence I didn't vouch for it. You'll have Windows folks coming to you: "bruh, because of your recent update I accidently shipped static libraries instead of shared." but that person would have to be quite dumb to do that anyway, but yea. 🤷 
I think your existing scripts are doing it right for the most part, just needed a little bit more touch and hence my PR. I honestly don't care if you end up merging it or not, it doesn't change anything for me nor my pride, rather just giving back to the community. I anyway use my own fork of your project or downstream project as you name it.

> I need to look at how other projects are handling the same issue. A new CMake variable may be required, but it would be a lot easier to support a binary variable than a variable with an arbitrary value.

From what I've seen so far, after building multiple big projects from source, including LLVM and many others, there are two common ways to approache this:
1) Build either static or shared library, don't allow both, then enforce the names to always be the same (i.e. `jpeg.lib`) <- I dislike this, because one may want to build both and that means it will require configuring cmake twice and build twice, meaning slower build time. Here is an example: https://github.com/gabime/spdlog/blob/v1.14.1/CMakeLists.txt#L155-L171
2) Allow building both shared and static (as your scripts currently do), but take good care for Windows folks, because Microsoft devs thought it would be a genious idea making shared and static libs using the same *.lib extension

I lean more towards solution 2), because it generates all the projects in one go and one can decide which target they want to build (i.e `cmake --build . --target jpeg-static` or `jpeg` for shared).

**Comment by dcommander** [comment]
I am out of the office and unable to thoroughly review everything you've submitted, but I will do so next week and make specific comments. I have not misunderstood, nor am I ignoring, anything. I was just trying to respond to you, as best I could, using my phone. I assure you that I will apply the full weight of my 15 years of experience with this code base when considering your patches.

**Comment by dcommander** [comment]
I can comment specifically on your proposals above. (1) is a non-starter, because our own packaging deploys both static and shared libraries. In (2), what do you mean by "take good care of Windows folks?" Is there some specific issue that would necessitate renaming our static libraries, other than awkward nomenclature? By the time libjpeg-turbo became a project, **jpeg.lib** was already the expected name for the libjpeg DLL import library, so I couldn't use that name for the static lib. I added a "-static" suffix because it seemed like the most straightforward approach. Most Windows developers just install our pre-built SDKs, so they are used to the "-static" suffix. Other than allowing consumers of the library to rename it to **jpeg.lib** if `ENABLE_SHARED` is off, I can't think of a less awkward/more straightforward approach than what we are currently doing. If `ENABLE_SHARED` is on, then we would have to use a suffix for the static lib regardless. What I chose to do is always use that suffix with the static libs, regardless of the value of `ENABLE_SHARED`, so consumers always know that they are static and not import libraries. That is a justifiable approach, even if it isn't your personal preference. Also, I'm pretty sure I got that idea from another project. (It may have been from one of the OpenSSL distributions. I can't recall.) Again, I will consider this more thoroughly next week. I don't need any more arguments, but if you have further research to share, such as specifics of how other projects deal with this naming conflict, then share them.

**Comment by dcommander** [comment]
Comments on specific patches:

https://github.com/libjpeg-turbo/libjpeg-turbo/pull/770/commits/03a622270ecddc93225156bc20196ff4c1475630
Rejected.    I understand why you prefer not to name the Windows static libraries **jpeg-static.lib** and **turbojpeg-static.lib**, since you aren't building shared libraries.  However:

1. Your case is not the common case.  The common case is distributing both static and shared libraries for Windows, in which case it is necessary to use a different name for the static library and the DLL import library.
2. By the time libjpeg-turbo was founded, other Win32 libjpeg API implementations (e.g. GnuWin) had already established **jpeg.lib** as a pseudo-standard name for the DLL import library. I chose to conform to that pseudo-standard, which precluded using **jpeg.lib** as the name for the static library.  (Also, most Win32 open source distributions tend to use a similar pseudo-standard.  Because DLL linking is more common than static linking, the DLL import library has naming priority.)
3. Referring to **BUILDING.md**, the use of a **-static** suffix for the Windows static libraries is documented behavior of libjpeg-turbo, and that behavior has been consistent since the project's inception 14 years ago.
4. The use of a **-static** suffix for the Windows static libraries is consistent with the behavior of other Windows binary distributions of open source libraries, such as [Win32OpenSSL](https://slproweb.com/products/Win32OpenSSL.html).  (In fact, Win32OpenSSL was probably my inspiration for doing similarly in libjpeg-turbo.)
5. The use of a **-static** suffix may not be your personal preference, but it does not place any undue usage burdens upon your project or other downstream projects.  It is impossible for me to make libjpeg-turbo conform to everyone's personal preferences, because many of those preferences are conflicting, but even if it were possible, doing so would make libjpeg-turbo an unmaintainable mess.
6. I frankly want to discourage downstream projects from renaming the static libraries to **jpeg.lib** and **turbojpeg.lib**, because those names are already used upstream for the DLL import libraries.  (Bear in mind that "downstream projects" include O/S-specific distributions of libjpeg-turbo meant to be used by developers, not just applications and DLLs that use libjpeg-turbo internally.)
7. I understand that this patch is backward-compatible, but introducing a new CMake variable implies full support for that variable.  That means that I would have to:
   1. Document the variable and its correct usage in **BUILDING.md**.
   2. Deal with potential naming conflicts if both static and shared libraries are being built.  (Referring to (6) above, I am opposed to allowing the static suffix to be removed.  I am more receptive to allowing the suffix to be changed to an arbitrary non-empty string.)
    3. Ensure that the packaging system (including the CMake config files and Windows installer script) properly accommodates the new variable.
    4. Test, test, test.

Our project has limited funding right now, and this simply isn't a high enough priority to justify the labor necessary to support the feature as you implemented it.  Per above, I am more receptive to allowing the **-static** suffix to be changed to an arbitrary non-empty string, if that would be acceptable to you.  That would eliminate the need for me to manage naming conflicts between the static and DLL import libraries, and it would be relatively easy to document and test that more limited feature (whereas your solution is much more error-prone, since it allows the base name of the static libraries to be changed.)

https://github.com/libjpeg-turbo/libjpeg-turbo/pull/770/commits/0f718795d47dfe21200a245328e8771c178bfc34
Rejected.  This patch represents a downstream personal preference imposed upon the upstream project.  If I introduced a static library suffix variable, per above, then that would presumably allow you to implement the same functionality that this patch implements, since you could change the static library suffix to `d` for `Debug` builds, `rd` for `RelWithDebInfo` builds, etc.  If I adopted per-configuration suffixes upstream, then I would have to document and thoroughly test that.  Technically it would be a backward-incompatible change, since the existing documented behavior is not to use per-configuration suffixes.

tl;dr: I am willing to add a CMake variable that allows you to change **-static** to some other non-empty string.  I am not willing to allow the static library base name to be changed, nor am I willing to allow the static library to be named the same as the DLL import library.

https://github.com/libjpeg-turbo/libjpeg-turbo/pull/770/commits/fbf911a34a58c41e1f11edbf51a5c26f52bb3097
Need more info.  It of course makes sense for libjpeg-turbo to install the linker PDB files for the DLLs, but in order to understand whether it makes sense to install the compiler PDB files for the static libraries, I need to understand the workflow in which those would be used.  Let's say a downstream project wants to link statically with a `Debug` build of libjpeg-turbo.  They would need to build libjpeg-turbo from source, since we don't distribute a `Debug` build of it upstream.  Let's assume they built libjpeg-turbo from source using the `Debug` configuration and `ENABLE_SHARED=0`, installed libjpeg-turbo into **c:\foo\lib**, built their downstream application/DLL using the `Debug` configuration, and linked with **c:\foo\lib\jpeg-static.lib**.  Would it be necessary for **c:\foo\lib\jpeg-static.pdb** to exist in order for the libjpeg-turbo debug symbols to be imported into the downstream application/DLL?  If so, then yes, I agree that we need to install the compiler PDB files for the static libraries.  However, I would need to change your patch slightly, since it doesn't currently accommodate config-less CMake generators (such as Ninja.)

**Comment by dcommander** [comment]
> [0f71879](https://github.com/libjpeg-turbo/libjpeg-turbo/commit/0f718795d47dfe21200a245328e8771c178bfc34) Rejected. This patch represents a downstream personal preference imposed upon the upstream project. If I introduced a static library suffix variable, per above, then that would presumably allow you to implement the same functionality that this patch implements, since you could change the static library suffix to `d` for `Debug` builds, `rd` for `RelWithDebInfo` builds, etc. If I adopted per-configuration suffixes upstream, then I would have to document and thoroughly test that. Technically it would be a backward-incompatible change, since the existing documented behavior is not to use per-configuration suffixes.

I realize that you are likely using a multi-config generator (Visual Studio), in which case a single static library suffix wouldn't allow you to exactly reproduce what this patch does.  The idea is that, if you are installing a single configuration of libjpeg-turbo, then a single static library suffix would be sufficient to reproduce the ultimate effect of the patch.

**Comment by dcommander** [comment]
Closing due to lack of response to my outstanding questions and proposals.
